Parallel circuits

SamWGran opened the pull request:
closes #13

SamWGran committed:
Added script files

WIP

SamWGran committed:
Work on Splitter logic

Added the Check functions for the splitter, along with return values to the PowerSource check. Next up is ResolveSplitter()

SamWGran committed:
Merge branch 'main' into parallel-circuits

SamWGran committed:
ResolveSplitter functions implemented

Not yet tested in any shape except for
on paper

SamWGran committed:
Splitter functionality

Added splitter checks and resolves to the battery functions.

SamWGran committed:
Added splitter script to prefab

SamWGran committed:
Small edits

Added script to splitter prefab, and commented some code.

SamWGran committed:
No Batteries in splitters

Added breaks if a battery is found inside of a splitter.

SamWGran committed:
Comments and small fixes

SamWGran committed:
Merge branch 'main' into parallel-circuits

steenish commented on Connected/Assets/Scripts/Components/Battery.cs:
```diff
@@ -29,27 +29,36 @@ private void OnDestroy()
+                    resistanceSum += splitterResistance;
+                }
+            } else if (nextComponent.GetType() == typeof(Combiner)) { // if battery is inside of a splitter-combiner.
+                //TODO: Add appropriate error message.
```
Do this!

steenish commented on Connected/Assets/Scripts/Components/Splitter.cs:
```diff
@@ -0,0 +1,84 @@
+
+        // If combiner1 isnt null and combiner2 is null it will catch after or sign.
+        if (combiner1 == null || (combiner1 != combiner2)) {
+            // TODO: Add appropriate error message
```
Do this!

steenish commented on Connected/Assets/Scripts/Components/Splitter.cs:
```diff
@@ -0,0 +1,84 @@
+                (res, nextComponent) = splitter.CheckSplitter();
+                resistanceSum += res;
+            } else if (nextComponent.GetType() == typeof(Battery)) {
+                //TODO: Add appropriate error message.
```
Do this!

SamWGran replied:
This is left undone due to us not having a completed way of giving the user text prompts. It is out of the scope of this branch so I left it like this

raaavioli replied:
Couldn't it be nice with some debug error messages for our own sake in case we hit those cases, and extend proper error messages for users later?

steenish commented on Connected/Assets/Scripts/Components/Splitter.cs:
```diff
@@ -0,0 +1,84 @@
+        // If combiner1 isnt null and combiner2 is null it will catch after or sign.
+        if (combiner1 == null || (combiner1 != combiner2)) {
+            // TODO: Add appropriate error message
+            return (0, null);
```
This tuple is used in a lot of different places, maybe we could make it a private static constant in this class?

SamWGran replied:
Will fix!

steenish reviewed (changes requested):
Shouldn't the splitter prefab have slots on it?

There is still a need to implement back-checking when a circuit is not closed, to reset the currents. I suspect that with the introduction of splitters and combiners, the back-checking code does not work. The splitter and combiner class will probably have to have methods ResetSplitter(), ResetSplitWire() and ResetCombiner().

raaavioli commented on Connected/Assets/Scripts/Components/Combiner.cs:
```diff
@@ -0,0 +1,18 @@
+
+public class Combiner : MonoBehaviour
+{
+    // Start is called before the first frame update
```
Shouldn't combiner have two negative wires, like Splitter has a second positive wire? 

I reckon that would be needed to be able to backtrack when breaking a circuit.

SamWGran replied:
Good catch! Combiner wasn't even a component. Will fix

SamWGran committed:
PR requests added

Debug logs added instead of comments, a constant error tuple was added and combiner.cs was corrected.

steenish reviewed (approved):
lgtm!
